DESIGN.md

```diff
@@ -0,0 +1,818 @@
+
+A group of experiments produces a _result_.
+
+Every result contains one or more _measurements_. The websites
+group produce a measurement for every URL that was tested. All the
+other groups produce a measurement for every experiment within
+the group itself. For example, the performance group currently is
+producing two measurements by default, generated by the NDT and
+DASH experiments.
+
+## Use cases
+
+We address two use cases: supporting the desktop app and running
+the CLI as an interactive tool.
+
+### Supporting the desktop app
+
+Because we want the desktop app to provide the same user experience
+of the mobile app, we need this functionality:
+
+1. discover the user IP, country, and network
+
+2. list the results stored on disk
+
+3. delete one or more results
+
+4. run any group of experiments (websites, instant messaging,
+performance, middleboxes), or run them all
+
+5. show each individual measurement
+
+6. upload measurements that were not uploaded
+
+To fully support this use case, we need to force the CLI to emit
+messages in JSON, which is easily parseable from Electron.
+
+### Interactive command line usage
+
+We also want users to run the CLI interactively. Users must be able
+to perform all the batch operations described above from the CLI with
+a reasonably pleasant and user friendly output. We will not specify
+exactly how this output is going to look like and we may change it
+without notice. Users that want to parse OONI output from scripts are
+encouraged to use the batch interface.
+
+## Software architecture
+
+The following diagram summarizes the software architecture:
+
+```
++--------------+   calls with flags   +----------+  .---------.
+| OONI Desktop | -------------------> | OONI CLI | | local DB |
++--------------+                      +----------+  `---------'
+          `--------<--<--<--<-------------'
+                   JSONL messages
+```
+
+The OONI CLI is a standalone, as static as possible, Go binary
+that allows OONI Desktop to perform all the actions defined
+above. The desktop app will communicate with the CLI using command
+line flags. The OONI CLI will send _on the standard ouput_ a
+stream of JSONL messages (i.e. JSON messages each terminated by
+a newline). In principle, we could emit any kind of JSON message,
+even though in practice, we emit JSON objects in most cases.
+
+The OONI CLI will statically link to Measurement Kit and/or
+ooni/probe-engine, and/or any other library required to
+perform OONI measurements.
+
+The OONI CLI will be updated by the desktop app.
+
+The OONI CLI will automatically download and store into
+a configuration/state directory any resource that it
+may require to perform its operations. This directory
+is by default `$HOME/.ooni` on all systems; yet, you can
+override it using the `OONI_HOME` environment variable.
+In the following we will call this directory the OONI home,
+state/configuration directory, or, with a stretch of
+semantics, `OONI_HOME`.
```

Comment:
I would add here that the CLI application needs to take into account the fact that a OONI Probe legacy installation MAY exist at this path and it needs to handle this case (it's already handled with a prompt that asks the user to either delete it or make a backup of it).

It's also worth noting that in the desktop installation this is set to another path specific for the desktop app, namely `~/Library/Application\ Support/OONI\ Probe/ooni_home/` on macOS and `XXX find what it is on windows` on windows

Reply:
> I would add here that the CLI application needs to take into account the fact that a OONI Probe legacy installation MAY exist at this path and it needs to handle this case (it's already handled with a prompt that asks the user to either delete it or make a backup of it).

On second thought, we should not be breaking existing installations. A better usage of our time seems to use a different directory. I would also give the new binary a different name, e.g., `ooniprobe3`, and just leave legacy installations alone. I bet people that has a working installation of ooniprobe legacy and want to just try out `ooniprobe3` to give us suggestions as it evolves would be very pleased by this arrangement. Also, this allows us to tell people to keep them back to back until we have replaced all features (i.e. less pressure on us).

> It's also worth noting that in the desktop installation this is set to another path specific for the desktop app, namely ~/Library/Application\ Support/OONI\ Probe/ooni_home/ on macOS and XXX find what it is on windows on windows

This is an implementation detail of the probe-desktop, I won't add it.